curl/src/main/scala/org/http4s/curl/websocket/CurlWSClient.scala

```diff
@@ -0,0 +1,207 @@
+import org.http4s.curl.internal.Utils
+import org.http4s.curl.internal.Utils.throwOnError
+import org.http4s.curl.unsafe.CurlExecutorScheduler
+import org.http4s.curl.unsafe.CurlRuntime
+import org.http4s.curl.unsafe.libcurl
+import org.http4s.curl.unsafe.libcurl_const
+import scodec.bits.ByteVector
+
+import scala.annotation.unused
+import scala.scalanative.unsafe._
+
+private[curl] object CurlWSClient {
+
+  final private val ws = Uri.Scheme.unsafeFromString("ws")
+  final private val wss = Uri.Scheme.unsafeFromString("wss")
+
+  private def setup(req: WSRequest, ec: CurlExecutorScheduler, verbose: Boolean)(
+      con: Connection
+  )(implicit zone: Zone) = IO {
+    val scheme = req.uri.scheme.getOrElse(ws)
+
+    if (scheme != ws && scheme != wss)
+      throw new RuntimeException(s"Websocket client can't handle ${scheme.value} scheme!")
+
+    val uri = req.uri.copy(scheme = Some(scheme))
+
+    throwOnError(
+      libcurl.curl_easy_setopt_customrequest(
+        con.handler,
+        libcurl_const.CURLOPT_CUSTOMREQUEST,
+        toCString(req.method.renderString),
+      )
+    )
+
+    if (verbose)
+      throwOnError(
+        libcurl.curl_easy_setopt_verbose(
+          con.handler,
+          libcurl_const.CURLOPT_VERBOSE,
+          1L,
+        )
+      )
+
+    // NOTE raw mode in curl needs decoding metadata in client side
+    // which is not implemented! so this client never receives a ping
+    // as all pings are always handled by libcurl itself
+    // throwOnError(
+    //   libcurl.curl_easy_setopt_websocket(
+    //     con.handler,
+    //     libcurl_const.CURLOPT_WS_OPTIONS,
+    //     libcurl_const.CURLWS_RAW_MODE,
+    //   )
+    // )
+
+    throwOnError(
+      libcurl.curl_easy_setopt_url(
+        con.handler,
+        libcurl_const.CURLOPT_URL,
+        toCString(uri.renderString),
+      )
+    )
+
+    // NOTE there is no need to handle object lifetime here,
+    // as Connection class and curl handler have the same lifetime
+    throwOnError {
+      libcurl.curl_easy_setopt_writedata(
+        con.handler,
+        libcurl_const.CURLOPT_WRITEDATA,
+        Utils.toPtr(con),
+      )
+    }
+
+    throwOnError {
+      libcurl.curl_easy_setopt_writefunction(
+        con.handler,
+        libcurl_const.CURLOPT_WRITEFUNCTION,
+        recvCallback(_, _, _, _),
+      )
+    }
+
+    libcurl.curl_easy_setopt_headerdata(
+      con.handler,
+      libcurl_const.CURLOPT_HEADERDATA,
+      Utils.toPtr(con),
+    )
+
+    throwOnError {
+      libcurl.curl_easy_setopt_headerfunction(
+        con.handler,
+        libcurl_const.CURLOPT_HEADERFUNCTION,
+        headerCallback(_, _, _, _),
+      )
+    }
+
+    var headers: Ptr[libcurl.curl_slist] = null
+    req.headers
+      .foreach { header =>
+        headers = libcurl.curl_slist_append(headers, toCString(header.toString))
+      }
+    throwOnError(
+      libcurl.curl_easy_setopt_httpheader(con.handler, libcurl_const.CURLOPT_HTTPHEADER, headers)
+    )
```

Comment:
Are these being freed somewhere? The doc says this
> When this option is passed to [curl_easy_setopt](https://curl.se/libcurl/c/curl_easy_setopt.html), libcurl will not copy the entire list so you must keep it around until you no longer use this handle for a transfer before you call [curl_slist_free_all](https://curl.se/libcurl/c/curl_slist_free_all.html) on the list.

Reply:
Oops! No, I thought that current zone is handling that, but apparently those are raw pointer created outside.

Reply:
I open a new issue to resolve this one, as curl client has a similar issue and this pull request is way to huge by now!